source/agora/node/QRCodeInterface.d

```diff
@@ -0,0 +1,278 @@
+/*******************************************************************************
+
+    Provides QR service interface used by the node's web-based management
+    interface
+
+    Using the web interface. This interface is provided by Vibe.d and
+    may be integrated into the node's web-based management in the future.
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.QRCodeInterface;
+
+import agora.common.crypto.Key;
+import agora.common.Hash;
+import agora.common.Types;
+import agora.consensus.data.PreImageInfo;
+
+import agora.utils.Log;
+import agora.node.data.LoginInfo;
+import agora.node.data.EncryptionKey;
+import agora.network.Clock;
+
+import vibe.core.core;
+import vibe.data.json;
+import vibe.http.router;
+import vibe.http.server;
+
+import std.datetime.date : DateTime;
+import std.datetime.systime : SysTime, Clock;
+import std.datetime.timezone : UTC;
+
+import core.time : days;
+
+import barcode;
+
+mixin AddLogger!();
+
+/*******************************************************************************
+
+    Class for Agora User's credentials QR services Interface
+
+    A class to provide a login or encryption key to QRs to verify that
+    the owner of the validator from outside agora.
+
+    This interface listens on 127.0.0.1:2829.
+    Respond SVG format's QR according to the get requests.
+
+    This interface must be in a path accessible only to the owner of the node.
+
+*******************************************************************************/
+
+public class QRCodeInterface
+{
+    /// HTTP listener
+    private HTTPListener listener;
+
+    /// The keypair of this node
+    private KeyPair key_pair;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            key_pair = the keypair of this node
+
+    ***************************************************************************/
+
+    public this (const KeyPair key_pair)
+    in
+    {
+        assert(KeyPair.init != key_pair);
+    }
+    do
+    {
+        this.key_pair = key_pair;
+    }
+
+    /***************************************************************************
+
+        Start listening for requests
+
+    ***************************************************************************/
+
+    public void start ()
+    {
+        auto settings = new HTTPServerSettings("127.0.0.1");
+        settings.port = 2829;
+        auto router = new URLRouter;
+        router
+            .get("/loginQR", &this.loginQRCreate)
+            .get("/encryptionKeyQR", &this.keyQRCreate);
+        this.listener = listenHTTP(settings, router);
+        scope (exit) this.listener.stopListening();
```

Comment:
That looks wrong. `listenHTTP` doesn't block, so  we `stopListening` immediately.